apps/server/src/modules/files-storage/service/preview.service.ts

```diff
@@ -0,0 +1,154 @@
+
+		return response;
+	}
+
+	private checkIfPreviewPossible(fileRecord: FileRecord): void | NotAcceptableException {
+		if (!fileRecord.isPreviewPossible()) {
+			this.logger.warn(`could not generate preview for : ${fileRecord.id} ${fileRecord.mimeType}`);
+			throw new UnprocessableEntityException(ErrorType.PREVIEW_NOT_POSSIBLE);
+		}
+	}
+
+	private async tryGetPreviewOrGenerate(params: PreviewFileParams): Promise<IGetFile> {
+		const { filePath, bytesRange } = params;
+		let file: IGetFile;
+
+		try {
+			file = await this.storageClient.get(filePath, bytesRange);
+		} catch (error) {
+			if (!(error instanceof NotFoundException)) {
+				throw error;
+			}
+
+			file = await this.generatePreview(params);
+		}
+
+		return file;
+	}
+
+	private createNameHash(params: DownloadFileParams, previewParams: PreviewParams): string {
+		const width = previewParams.width ?? '';
+		const height = previewParams.height ?? '';
+		const format = previewParams.outputFormat ?? '';
+		const fileParamsString = `${params.fileRecordId}${width}${height}${format}`;
+		const hash = crypto.createHash('md5').update(fileParamsString).digest('hex');
+
+		return hash;
+	}
+
+	private async generatePreview(params: PreviewFileParams): Promise<IGetFile> {
+		const { fileRecord, downloadParams, previewParams, hash, filePath, bytesRange } = params;
+
+		const original = await this.fileStorageService.download(fileRecord, downloadParams, bytesRange);
+		const preview = this.resizeAndConvert(original, fileRecord, previewParams);
+
+		const format = previewParams.outputFormat ?? fileRecord.mimeType;
+		const fileDto = FileDtoBuilder.build(hash, preview, format);
+		await this.storageClient.create(filePath, fileDto);
+
+		const response = await this.storageClient.get(filePath, bytesRange);
+
+		return response;
+	}
+
+	private resizeAndConvert(
+		original: IGetFileResponse,
+		fileRecord: FileRecord,
+		previewParams: PreviewParams
+	): PassThrough {
+		const mimeType = previewParams.outputFormat ?? fileRecord.mimeType;
+		const format = this.getFormat(mimeType);
+		const im = subClass({ imageMagick: true });
+
+		const preview = im(original.data, fileRecord.name);
+		const { width, height } = previewParams;
+
+		if (width) {
+			preview.resize(width, height);
+		}
+
+		const result = preview.stream(format);
+
+		return result;
+	}
+
+	private getFormat(mimeType: string): string {
+		const format = mimeType.split('/')[1];
+
+		return format;
+	}
+
+	private getPreviewName(fileRecord: FileRecord, outputFormat?: PreviewOutputMimeTypes): string {
+		if (!outputFormat) {
+			return fileRecord.name;
+		}
+
+		const fileNameWithoutExtension = fileRecord.name.split('.')[0];
+		const format = this.getFormat(outputFormat);
+		const name = `${fileNameWithoutExtension}.${format}`;
+
+		return name;
+	}
+
+	private getFilePath(fileRecord: FileRecord, hash: string): string {
+		const path = ['previews', fileRecord.getSchoolId(), hash].join('/');
```

Comment:
I thought about it again and I still think it would be a good idea to put the original filerecord id in the path too. Like:
```suggestion
		const path = ['previews', fileRecord.getSchoolId(), fileRecord.id, hash].join('/');
```
Apparently we don't want to delete the previews right away when the original file is deleted... at the moment. But this might change in the future. But also besides that, it would make it easy to find the previews for one file. And maybe sometime you want to do that. 

Does anything speak against doing it like I suggest?